pandas/core/indexes/base.py

```diff
@@ -3718,9 +3719,17 @@ def get_indexer(
 
             indexer = self._engine.get_indexer(target.codes)
             if self.hasnans and target.hasnans:
+                # After _maybe_cast_listlike_indexer, target elements which do not
+                # belong to some category are changed to NaNs
+                # Mask to track actual NaN values compared to inserted NaN values
+                try:
+                    target_nans = isna(orig_target)
+                except NotImplementedError:
+                    target_nans = False
```

Comment:
is this ever reached?

Reply:
Yes, whenever ci has NaNs, and target has elements which do not map to an existing category

Reply:
why would that raise NotImplementedError?

Reply:
Yes, it seems that this error won't be raised. But wouldn't it be better if the try-except remains in place?

Reply:
> But wouldn't it be better if the try-except remains in place?

Well, no.  If an exception is raised by isna then we should raise to let the user know about it.

Reply:
Yeah, that makes sense. I've made the necessary changes,